earth_enterprise/src/common/geFilePool.cpp

```diff
@@ -20,17 +20,65 @@
 #include <third_party/rsa_md5/crc32.h>
 #include <khEndian.h>
 
+// ****************************************************************************
+// ***  POSIXFileAcessor
+// ****************************************************************************
+int POSIXFileAccessor::Open(const std::string &fname, int flags, mode_t createMask) {
+  return khOpen(fname, flags, createMask);
+}
+
+int POSIXFileAccessor::FsyncAndClose(std::shared_ptr<AbstractFileIdentifier> pID) {
+  return khFsyncAndClose(pID->getAsFD());
+}
+
+int POSIXFileAccessor::Close(std::shared_ptr<AbstractFileIdentifier> pID) {
+  return khClose(pID->getAsFD());
+}
+
+bool POSIXFileAccessor::PreadAll(std::shared_ptr<AbstractFileIdentifier> pID, void* buffer, size_t size, off64_t offset) {
+  return khPreadAll(pID->getAsFD(), buffer, size, offset);
+}
+
+bool POSIXFileAccessor::PwriteAll(std::shared_ptr<AbstractFileIdentifier> pID, const void* buffer, size_t size, off64_t offset) {
+  return khPwriteAll(pID->getAsFD(), buffer, size, offset);
+}
+
+// ****************************************************************************
+// ***  FileIdentifierFactory
+// ****************************************************************************
+namespace FileIdentifierFactory {
+  static std::shared_ptr<AbstractFileIdentifier> getIdentifier(int i) {
+    return std::make_shared<POSIXFileIdentifier>(i);
+  }
+}
+
+// ****************************************************************************
+// ***  FileAccessorFactory
+// ****************************************************************************
+namespace FileAccessorFactory {
+  static std::shared_ptr<AbstractFileAccessor> getAccessor(const std::string &fname) {
+    return std::make_shared<POSIXFileAccessor>();
+  }
+
+  static std::shared_ptr<AbstractFileAccessor> getAccessor(std::shared_ptr<AbstractFileIdentifier> aID) {
+    if ( typeid(*aID) == typeid(POSIXFileIdentifier) ) {
+      return std::make_shared<POSIXFileAccessor>();
+    }
+    return nullptr;
+  }
+}
+
 // ****************************************************************************
 // ***  FileReservationImpl
 // ****************************************************************************
 bool FileReservationImpl::UnlockAndClose_(geFilePool &pool) {
   // run when pool.mutex IS locked
-  if (fd != -1) {
+  if (fid->isValid()) {
     int result = 0;
     {
       khUnlockGuard unlock(pool.mutex);
-      result = isWriter ? khFsyncAndClose(fd) : khClose(fd);
-      fd = -1;
+      result = isWriter ? FileAccessorFactory::getAccessor(Fid())->FsyncAndClose(Fid()) : FileAccessorFactory::getAccessor(Fid())->Close(Fid());
```

Comment:
Minor: you could make this cleaner by calling `getAccessor` once (maybe in the constructor) and storing a pointer to it.